File: Programming_2/C++/lab_5/lab_5.1/Trapezoid.cpp
//
// Created by Danil on 23.03.2020.
//
#include <cmath>
#include <ostream>
#include "Trapezoid.h"

Trapezoid::Trapezoid(const Point *argv): Figure(size, argv) {}

double Trapezoid::Area() const {
    return  (lenSide(points[0], points[3]) + lenSide(points[1], points[2])) *
    std::fabs(points[1].get_y() - points[0].get_y()) / 2;
}

double Trapezoid::Perimeter() const {
    return round(lenSide(points[0], points[1])
                      + lenSide(points[1], points[2])
                      + lenSide(points[2], points[3])
                      + lenSide(points[0], points[3]));
}

std::ostream &operator<<(std::ostream &stream, const Trapezoid &trapezoid) {
    for (int i = 0; i < Trapezoid::size; i++)
    {
        stream << "Point" << i+1 << ": " << trapezoid.points[i] << std::endl;
    }
    stream << "\nArea: " << trapezoid.Area() << "\nPerimeter: " << trapezoid.Perimeter() << std::endl;
    return stream;
}

File: Programming_2/C#/lab_5/lab_5.1/Trapezoid.cs
using System;

namespace lab_5
{
    public class Trapezoid : Figure
    {
        public Trapezoid(params Point[] points):base(points) {}
        
        public double Area()
        {
            return (SideLength(Points[0], Points[3]) + SideLength(Points[1], Points[2])) *
                    Math.Abs(Points[1].Y - Points[0].Y) / 2;
        }
        
        public double Perimeter()
        {
            return Math.Round(SideLength(Points[0], Points[1])
                   + SideLength(Points[1], Points[2])
                   + SideLength(Points[2], Points[3]) 
                   + SideLength(Points[0], Points[3]), 3);
        }
        
        public override string ToString()
        {
            return $"Point:\nA{Points[0]}  B{Points[1]}  C{Points[2]}  D{Points[3]}\nArea: {Area()}   Perimeter: {Perimeter()}";
        }
    }
}